SRXL2 failsafe fix for newer receivers

SpektrumRC opened the pull request:
New firmware coming for the SPM4650 (and all future SRXL2 receivers) will alternate between dBm (negative) and percent (non-negative) RSSI values with each ChannelData packet to allow either type of value to be captured. Since betaflight only displays %, dBm should just be ignored to avoid misleading average values.
This change in betaflight SRXL2 code addresses failsafe state not being detected properly by the alternating RSSI values.

etracer65 commented:
This seems like a problematic implementation in the receiver firmware. You're opening the door for anybody who has updated their receiver firmware yet using any Betaflight version older than now (or even other firmwares) to have a non-functional failsafe.

mikeller commented:
@SpektrumRC: I agree with @etracer65 in that I do not think that this firmware update for your RX can be released in this form - failsafe will stop working for anybody using it with the currently released version of Betaflight, and this is a risk that should not be taken.

The reason for this is that failsafe detection requires the failsafe status to be present for a minimum amount of time before failsafe stage 2 is activated, but the way failsafe is detected in the SRXL2 code in Betaflight up to and including 4.1 means that this failsafe will only ever be detected intermittently on every second frame with the new SRXL2 firmware.

As a mitigation, can I propose that the implementation of the SRXL2 firmware is changed so that in the case of a failsafe both the RSSI and the dBm slots send 0 as a value - this will maintain compatibility with the current implementation, the 'drawback' of not getting dBm values during a failsafe event seems to be minor drawback compared to the prospect of craft flying away over the horizon because failsafe does not work any more.

This change will mean that the `|| (channelData->rssi < 0 && lastRssi == 0))` part of the check for failsafe can be removed.

SpektrumRC commented:
@etracer65  & @mikeller : We agree that this is somewhat problematic, and wish we had fixed this before the 4.1 release. Unfortunately, this change in SRXL2 remote receiver RSSI behavior is required to solve some more fundamental issues across multiple products and multiple use cases beyond betaflight flight controllers. In other words, it's not as simple as releasing firmware that sends 0 in both % and dBm slots.

To be clear, this only impacts betaflight since it is caused by the decision to use RSSI == 0 as the basis for entering failsafe mode. This was done to give fine control over failsafe timing rather than depending on the receiver failsafe state. In hindsight, we should have used either condition to enter failsafe mode. The new receiver itself will still cut throttle when it enters our failsafe state, but will remain armed without this patch. There is no risk of the drone flying away over the horizon -- it still falls to the ground.

I should also clarify that we have done considerable testing and found that this patch fully resolves this issue, and still works fine with old receiver firmware. So the only circumstances in which this is a concern would be the scenario where someone is already using an old betaflight version without this patch, but they replace their receiver with an SPM4651T or newer receiver without updating their betaflight version.

To mitigate risk we do not intend to update the firmware for the SPM4650 any time soon -- this will remain the same for the time being on factory programmed devices. We are also going to have prominent notices  and markings on the affected products and product documentation that points to the required betaflight build for proper failsafes function.

mikeller commented:
@SpektrumRC:

> Unfortunately, this change in SRXL2 remote receiver RSSI behavior is required to solve some more fundamental issues across multiple products and multiple use cases beyond betaflight flight controllers. In other words, it's not as simple as releasing firmware that sends 0 in both % and dBm slots.

Can you explain why the minor changes of sending 0 instead of the dBm value on the RX => flight controller interface in the case of a failsafe is problematic?

etracer65 commented:
> The new receiver itself will still cut throttle when it enters our failsafe state, but will remain armed without this patch. There is no risk of the drone flying away over the horizon -- it still falls to the ground.

This is actually far more dangerous. The receiver will cut the throttle channel to zero, so? That doesn't stop the motors from spinning. The quad is going to fall to the ground fully armed with the motors spinning and PID controller active looking to kill anybody it can find. At the very least it's going to destroy the quad and burn out motors and ESCs. Even in the unlikely case that the pilot is using MOTOR_STOP (which nobody does as it interferes with flight characteristics), then there's the risk that as the pilot walks forward to retrieve their crashed quad that the motors might suddenly start again as signal is restored. Once again, on the ground flipping out, destroying the quad, or worse. The pilot's buddy walks out to pick up the quad and is walking back. Once he gets back in range the motors start - bad day for that person.

Also this will make your receiver unusable for many users. Since the flight controller will have no knowledge of the failsafe there's no way to make use of failsafe RTH capability like GPS Rescue in Betaflight or other RTH capabilities in firmwares like iNav.

etracer65 commented:
> The new receiver itself will still cut throttle when it enters our failsafe state

And how would you know what channel the user has set up for throttle? I assume you are talking about the standard channel mapping for Spektrum but there's no requirement that the user has mapped throttle to that channel.

SpektrumRC commented:
@mikeller & @etracer65: We hear your concerns, and take them very seriously. The main obstacle for us is retesting/recertifying receiver firmware changes while managing supply chain and production in a very short timespan. These actions can negatively impact multiple other products. We are currently expediting a firmware build that would set RSSI to 0 when the receiver enters failsafe, which addresses your safety concerns by disarming old Betaflight releases when the receiver enters failsafe. We don't expect problems with this change and will follow up accordingly. 

Regardless of our firmware development, we want to ensure that the necessary changes in this pull request are made. We feel that it is very important for Betaflight failsafe function to be robust and properly supported with all future SRXL2 products. We will ensure that Spektrum firmware follows suit including the SPM4650 and apologize for the headaches created in these changes. 

Notably we want to suggest that we do not remove the code that @mikeller suggested because that would mean that the flight controller would not enter failsafe until the receiver detects the failsafe, which takes about 1 second. That code ensures that RSSI == 0 during runtime is still used to trigger the failsafe after your timeout, which iirc is more like 150ms. We understand this to be the desired Betaflight behavior.

mikeller commented:
@SpektrumRC: Hit me up in [Slack](https://slack.betaflight.com/).

SpektrumRC commented:
I had some discussion with @mikeller, and we've settled on using the empty channel mask as a way of detecting packet losses while the receiver is still in a "connected" state (based on our definition of connected vs failsafe). When the receiver is in it's own failsafe state, the channel mask will no longer be empty, but betaflight will still flag the packet as a failsafe since the channel data command from the receiver changes to "FailsafeChannelData" rather than just "ChannelData".
In addition to that from some of the previous discussion, the new receivers/firmware will send 0 rssi on a "receiver failsafe", so older betaflight firmware will still end up entering a failsafe with newer receiver firmware. I've measured this out to be approx 1380ms total from the moment RF signal stops, to the motors being disarmed. With the lastest of both receiver firmware and betaflight firmware, the delay to disarm is approx 870ms.

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -168,18 +168,17 @@ bool srxl2ProcessHandshake(const Srxl2Header* header)
-        globalResult = RX_FRAME_FAILSAFE;
-    } else {
-        globalResult = RX_FRAME_COMPLETE;
+    //If receiver is in a connected state, and a packet is missed, the channel mask will be 0.
```
This looks a lot better now, thanks for changing this @SpektrumRC!

mikeller commented on src/main/rx/srxl2.c:
```diff
@@ -210,6 +209,7 @@ bool srxl2ProcessControlData(const Srxl2Header* header, rxRuntimeState_t *rxRunt
 
     case FailsafeChannelData: {
         srxl2ProcessChannelData((const Srxl2ChannelDataHeader *) (controlData + 1), rxRuntimeState);
+        globalResult |= RX_FRAME_FAILSAFE;
```
Just a quick question: Why is `srxl2ProcessChannelData()` called above here? If the link is in a failsafe condition, there should be no new RC data, and Betaflight won't be using it anyway while in failsafe mode.

SpektrumRC replied:
It does not necessarily have to be called. Just makes the failsafe channel data available in case betaflight want's to use it. I have no problems removing it if betaflight has no future intention of using the receivers own failsafe channel values for any purpose.

SpektrumRC replied:
In terms of receiver behavior, the default is that throttle goes to a preset position, usually a low/idle value (whatever it was set to on the transmitter when it was bound to the receiver), and everything else stays at the last received position.  I say default, as we may give users the option to modify this for serial receivers in the future.

mikeller replied:
The expected RX behaviour in Betaflight is that in case of a failsafe the RX stops sending RC channel values, and does not repeat the last good values.
In cases like SRXL2 where the RX explicitly sets `RX_FRAME_FAILSAFE` this isn't strictly required, as channel values are ignored when this flag is set. But I think it will make the code easier to understand if the `srxl2ProcessChannelData()` inside `FailsafeChannelData` is removed.

etracer65 commented:
@SpektrumRC 

> the default is that throttle goes to a preset position, usually a low/idle value (whatever it was set to on the transmitter when it was bound to the receiver), and everything else stays at the last received position. I say default, as we may give users the option to modify this for serial receivers in the future.

When does this happen to the throttle? Immediately when the receiver detects a signal loss or when it tells Betaflight of the failsafe? If it's the former than that's a problem for anybody trying to use GPS Rescue or any other failsafe procedure as the quad will start to fall from the sky for that interim period. Most receivers hold ALL channels during this period (sometimes called "stage 0" failsafe).

SpektrumRC commented:
@etracer65 It is when the receiver itself has entered a failsafe state.
This is provided through a different SRXL2 command (FailsafeChannelData), so the FC has the choice of using that channel data or ignoring it. The GPS rescue is a valid concern in that case, and I dont have a GPS equipped quad to test if thats the case unfortunately. I went ahead and removed the channel processing on failsafe data either way to erase any doubts, and squashed the commits.
